p2p: Drop m_recently_announced_invs bloom filter

ajtowns opened the pull request:
This PR replaces the `m_recently_announced_invs` bloom filter with a simple sequence number tracking the mempool state when we last considered sending an INV message to a node. This saves 33kB per peer (or more if we raise the rate at which we relay transactions over the network, in which case we would need to increase the size of the bloom filter proportionally).

The philosophy here (compare with #18861 and #19109) is that we consider the rate limiting on INV messages to only be about saving bandwidth and not protecting privacy, and therefore after you receive an INV message, it's immediately fair game to request any transaction that was in the mempool at the time the INV message was sent. We likewise consider the BIP 133 feefilter and BIP 37 bloom filters to be bandwidth optimisations here, and treat transactions as requestable if they would have been announced without those filters. Given that philosophy, tracking the timestamp of the last INV message and comparing that against the mempool entry time allows removal of each of `m_recently_announced_invs`, `m_last_mempool_req` and `UNCONDITIONAL_RELAY_DELAY` and associated logic.

ajtowns committed:
net_processing: drop m_recently_announced_invs bloom filter

Rather than using a bloom filter to track announced invs, simply allow
a peer to request any tx that entered the mempool prior to the last INV
message we sent them. This also obsoletes the UNCONDITIONAL_RELAY_DELAY.

ajtowns committed:
net_processing: Clean up INVENTORY_BROADCAST_MAX constants

ajtowns commented:
As justification for ignoring the privacy benefits of rate limiting INV messages, consider the following approach: an attacker makes multiple inbound connections to our peer, and upon receiving an INV of txs a,b,c,d,... via one connection immediately announces the same txs on each of its other connections. In that case the next INV we send on some other connection to the attacker will skip those transactions, providing a new unique set. With this approach, the attacker should be able to receive tx announcements from us at many times the 7tx/s limit we originally set.

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -126,7 +126,7 @@ class CTxMemPoolEntry
     }
     size_t GetTxWeight() const { return nTxWeight; }
-    std::chrono::seconds GetTime() const { return std::chrono::seconds{nTime}; }
+    NodeClock::time_point GetNodeTime() const { return nTime; }
```
I mostly renamed it to make sure there'd be compiler errors anywhere that it didn't get updated, just in case some type conversion happened automatically.

ajtowns replied:
This isn't renamed anymore, instead `GetEntryTime()` is added for the steady clock time tracking mempool entry. (Happy to rename it to something better if someone has a suggestion)

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2261,26 +2253,17 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
-        if (mi != mapRelay.end()) return mi->second;
-    }
+    // Otherwise, the transaction must have been announced recently or removed from the mempool.
+    // If the latter, maybe it's in mapRelay!
```
Yeah, this is broken as-is. My original idea was to add a timestamp to the `mapRelay` entries too, then it was to drop `mapRelay` and add `vExtraTxnForCompact`-relay, then I compromised by just doing it wrong...

ajtowns replied:
This includes commits to drop `mapRelay` now (cf #27625) so shouldn't be broken anymore.

sdaftuar commented:
Concept ACK on getting rid of the recently announced filter.  I think we'd need to use a steady clock to avoid random failures for no good reason, is that easy to do?

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2299,7 +2299,7 @@ void PeerManagerImpl::ProcessGetData(CNode& pfrom, Peer& peer, const std::atomic
     const CNetMsgMaker msgMaker(pfrom.GetCommonVersion());
 
-    const auto now{GetTime<std::chrono::seconds>()};
+    const NodeClock::time_point now = NodeClock::now();
```
We could; it'd mean adding an extra field to the mempool, and using both mockable time and steady time for dealing with next invs (mockable time so we can skip over delays in sending invs; steady time for this purpose).

ajtowns replied:
Switched to steady clock.

sr-gi commented:
What's the rationale regarding optimizing for memory instead of privacy here? I can see how this simplifies the design, however, I fail to see this being an obvious pick given we'd be potentially even more exposed to fingerprinting than before (as is now it is trivial for every connection to probe data on the node's mempool).

sr-gi commented:
So just to be clear, my concern here relates to the fact that the current codebase only allows a peer to request a transaction as long as we have announced that to them (with the exception of data requested after a `MEMPOOL` message if we happen run signaling bip37).

After this patch, the logic above won't hold true anymore. If for whatever reason we have more than `INV_BROADCAST_MAX` transactions to send to a given peer at a given time, and we send an INV message to it, it will immediately be able to request any of those new transactions (even if they were never offered). Whether this is in itself a meaningful privacy leak is unclear to me. Under regular loads (<`INV_BROADCAST_MAX` pending transactions to be announced), this behaves exactly like the current codebase.

I guess it all boils down to whether potentially allowing a peer to know when a transaction enters our mempool is something we should guard about or not.

ajtowns commented:
> What's the rationale regarding optimizing for memory instead of privacy here? I can see how this simplifies the design, however, I fail to see this being an obvious pick given we'd be potentially even more exposed to fingerprinting than before (as is now it is trivial for every connection to probe data on the node's mempool).

This isn't trading off privacy -- that's the point of [this comment](https://github.com/bitcoin/bitcoin/pull/27675#issuecomment-1549711784). It does mean that someone trying to find out whether a recent transaction entered your mempool in the last window doesn't need multiple inbound connections, but that's a good thing: having many inbound connections is easy for an attacker, so avoiding the need to make many inbound connections just means those inbound slots become available for honest peers.

> So just to be clear, my concern here relates to the fact that the current codebase only allows a peer to request a transaction as long as we have announced that to them (with the exception of data requested after a `MEMPOOL` message if we happen run signaling bip37).

That's already the case for any txs that have been in the mempool for greater than two minutes, and any transactions that have been announced to a peer. Generally that's plenty for fingerprinting -- if you announce pairs of conflicting txs randomly to different nodes, then if you do that 20 times at the same low-ish feerate (so that neither will rbf the other), then wait two minutes, you'll likely already get a nearly unique set of txs that will serve as a fingerprint for any given node.

> After this patch, the logic above won't hold true anymore. If for whatever reason we have more than `INV_BROADCAST_MAX` transactions to send to a given peer at a given time, and we send an INV message to it, it will immediately be able to request any of those new transactions (even if they were never offered). Whether this is in itself a meaningful privacy leak is unclear to me. Under regular loads (<`INV_BROADCAST_MAX` pending transactions to be announced), this behaves exactly like the current codebase.
> 
> I guess it all boils down to whether potentially allowing a peer to know when a transaction enters our mempool is something we should guard about or not.

The point of the random INV intervals is that we limit the knowledge of when a transaction entered our mempool to a specific window -- if we send an INV at t=0s, t=1.1s, t=6.1s, t=6.4s, eg, then our peer can tell the difference between a tx that entered at t=0.5s and t=3s, but can't tell the difference between t=3s and t=4s or t=5s, because they're all in the same window. We then use the same window boundaries for all inbound peers so that adding extra inbound peers doesn't give you more precise knowledge.

If the attacker only has a single inbound connection, then when the number of txs in a window exceeds `INV_BROADCAST_MAX` they'll get less information; however as the comment above attempted to explain, adding additional inbound connections and manually announcing some txs on some of the connections will allow an attacker to get the full set of txs included in each window. So I don't think it makes sense to consider the `INV_BROADCAST_MAX` limit to have a privacy benefit. (And if we eventually add erlay support, I think that limit goes away entirely for connections doing announcements over erlay rather than INV anyway)

sr-gi commented:
> > What's the rationale regarding optimizing for memory instead of privacy here? I can see how this simplifies the design, however, I fail to see this being an obvious pick given we'd be potentially even more exposed to fingerprinting than before (as is now it is trivial for every connection to probe data on the node's mempool).
> 
> This isn't trading off privacy -- that's the point of [this comment](https://github.com/bitcoin/bitcoin/pull/27675#issuecomment-1549711784). It does mean that someone trying to find out whether a recent transaction entered your mempool in the last window doesn't need multiple inbound connections, but that's a good thing: having many inbound connections is easy for an attacker, so avoiding the need to make many inbound connections just means those inbound slots become available for honest peers.

Right, I don't think I originally fully understood what you meant by that. So the case is:

- An attacker has `n` inbound connections to node `B` (namely `A0, ..., An-1`)
- `B` sends an `INV` message containing `a, b, c, d` through link `Ai`
    - *During the same message processing loop* for `B`, `A` sends `a, b, c, d` through link `Aj`
    - Now `Aj` will receive an `INV` with a disjoint set for `a, b, c, d` when `B` picks it in this round

Provided `A` has enough links with `B`, it should be able to get all pending connections in `B`'s mempool in a single round. Is that it? If so, I think this renders all my other comments irrelevant.

> That's already the case for any txs that have been in the mempool for greater than two minutes, and any transactions that have been announced to a peer. Generally that's plenty for fingerprinting -- if you announce pairs of conflicting txs randomly to different nodes, then if you do that 20 times at the same low-ish feerate (so that neither will rbf the other), then wait two minutes, you'll likely already get a nearly unique set of txs that will serve as a fingerprint for any given node.

I don't think I follow here though. Would you mind elaborating?

ajtowns commented on src/kernel/mempool_persist.cpp:
```diff
@@ -78,7 +78,7 @@ bool LoadMempool(CTxMemPool& pool, const fs::path& load_path, Chainstate& active
             if (nTime > TicksSinceEpoch<std::chrono::seconds>(now - pool.m_expiry)) {
                 LOCK(cs_main);
-                const auto& accepted = AcceptToMemoryPool(active_chainstate, tx, nTime, /*bypass_limits=*/false, /*test_accept=*/false);
+                const auto& accepted = AcceptToMemoryPool(active_chainstate, tx, NodeClock::time_point{std::chrono::seconds{nTime}}, /*bypass_limits=*/false, /*test_accept=*/false);
```
I think the answer is just to add a new tx entry field for the steady clock time and use that at natural precision.

ajtowns replied:
Entry time is no longer saved or loaded, so this is no longer relevant.

ajtowns committed:
mempool_entry: add mempool entry sequence number

ajtowns commented:
> > [it is trivial for every connection to probe data on the node's mempool] already the case for any txs that have been in the mempool for greater than two minutes, and any transactions that have been announced to a peer. Generally that's plenty for fingerprinting -- if you announce pairs of conflicting txs randomly to different nodes, then if you do that 20 times at the same low-ish feerate (so that neither will rbf the other), then wait two minutes, you'll likely already get a nearly unique set of txs that will serve as a fingerprint for any given node.
> 
> I don't think I follow here though. Would you mind elaborating?

The easy way to probe for the presence of transaction X in the mempool is to propose a fake transaction Z that spends an unspent output of X, and a fake transaction Y, with fake signatures for both, and see how its orphan handling behaves. If you receive a request for X and Y, then the victim didn't have X in its mempool; if you receive a request for just Y, then it did have X in the mempool; if you don't receive a request for either, then X was not in the mempool and was in its reject filter.

If you want to fingerprint a node, for example to see that an ipv4 address and a tor address are both pointing to the same node, then you need to set things up so that each node in the network has a different set of txs in their mempools, then query them to see if the node behind the ipv4 address and the node behind the tor address have the same set of txs. One way to do that is to propose a bunch of double-spends at the same feerate: a1/a2, b1/b2, c1/c2, etc and fire them off at various nodes on the network, then use the previous technique to see which txs actually ended up in the interesting nodes' mempools.

None of that requires careful timing -- you just need to set the feerates so they won't either be mined immediately, nor be evicted from the mempool too quickly.

ajtowns commented:
> I don't think the multiple inbound connections argument is convincing, unless I am missing something. In practise, the inbound connections are scheduled randomly, so the attacker wouldn't know which inbound is the next one to announce the known invs on. And once they were scheduled by the victim (or the first inv message was sent by the attacker), it is too late to send another inv message, because only one inv message is processed per scheduling round.

I think that's incorrect: if you have three inbound peers, A, B and C, the timing may look something like:

 * t=0, `next_inv_to_inbounds = 4.8s`
 * t=1, `SendMessages`, `peerB->m_next_inv_send_time = next_inv_to_inbounds`
 * t=1.001, `SendMessages`, `peerA->m_next_inv_send_time = next_inv_to_inbounds`
 * t=1.002, `SendMessages`, `peerC->m_next_inv_send_time = next_inv_to_inbounds`
 * t=2, `RelayTransaction(tx1)`
 * t=3, `RelayTransaction(tx2)`
 * t=4, `RelayTransaction(tx3)`
 * t=5, `SendMessages`, `peerC->m_next_inv_send_time = next_inv_to_inbounds = 10s`, INV tx1, tx2, tx3 to C
 * t=5.001, `SendMessages`, `peerB->m_next_inv_send_Time = next_inv_to_inbounds`, INV tx1, tx2, tx3 to B
 * t=5.002, `SendMessages`, `peerA->m_next_inv_send_Time = next_inv_to_inbounds`, INV tx1, tx2, tx3 to A

Note that `NextInvToInbounds` only gets a new time if `next_inv_to_inbounds < now`, which leaves plenty of time for all the inbound peers to be cycled through before a new time is picked.

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -127,6 +130,7 @@ class CTxMemPoolEntry
     }
     size_t GetTxWeight() const { return nTxWeight; }
     std::chrono::seconds GetTime() const { return std::chrono::seconds{nTime}; }
+    SteadyClock::time_point GetEntryTime() const { return m_entry_time; }
```
> Looks like this one is only to be used internally for relay and the other one is to be used for everything else?

Yeah - `SteadyClock` times aren't comparable between runs (particularly if there was a reboot inbetween), so these timestamps are only useful "internally" in general.

Could alternatively store the value of `mempool->GetSequence()` instead of the `SteadyClock` time for a similar result.

ajtowns replied:
Updated to use the mempool sequence number rather than `SteadyClock`

ajtowns commented:
> It looks like all inbound peers receive the same INVs, which doesn't look like an attack, but normal operation? 

Yes, that's ordinary operation, where everyone finds out all the txids you received in an interval automatically. If you'd instead received more than 35 txids in the interval (tx1 .. tx100, say, with tx1 being highest feerate (`compareInvMempoolOrder`)) then everyone would receive `INV tx1..tx35`. For an attacker who wants to know the timing information for all 100 txs, they could make 3 inbound connections, `peerA1` through `peerA3`, as follows:

 * t=4.5, `peerA1` sends INV tx1..tx66 to us
 * t=4.51, `peerA2` sends INV tx34..tx100 to us
 * t=4.52, `peer A3` sends INV tx1..tx33, tx67..tx100 to us

and our response would then be to remove the INVed txs from each peer's queue, so when we hit `next_inv_to_inbounds` we'd do:

 * t=5, `SendMessages`, `peerC->m_next_inv_send_time = next_inv_to_inbounds = 10s`, INV tx1..tx35 to C 
 * t=5.001, `SendMessages`, `peerA1->m_next_inv_send_time = next_inv_to_inbounds`, INV tx67..tx100 to A1
 * t=5.002, `SendMessages`, `peerA3->m_next_inv_send_time = next_inv_to_inbounds`, INV tx34..tx66 to A3
 * t=5.003, `SendMessages`, `peerA2->m_next_inv_send_time = next_inv_to_inbounds`, INV tx1..tx33 to A2
 * t=5.004, `SendMessages`, `peerB->m_next_inv_send_time = next_inv_to_inbounds`, INV tx1..tx35 to B

and between A1..A3 you receive announcements for all of tx1..tx100 that were accepted into the victim's mempool between the `next_inv_to_inbounds` timestamps of t=0 and t=4.8.

ie, basically if the network's processing up to 7X tx/second, you make X connections to the victim peer, divide the txs you receive into X groups, and announce all txids except for the K'th group to the victim via connection K. (If they request the tx from you, then just quickly reply notfound to avoid biasing the data). Few things to be careful of if you want to actually make that work, but should be find in principle.

ajtowns commented:
> Ok, looks like you are assuming that the attacker already knows all transactions (or txids), and I assumed this is what the attacker wants to find out.

You find out txids more quickly just by running a public listening node; the extra information that active probing gets you is information about when a tx was first received by different nodes in the network, from which you can then infer some idea about how the network is connected ("which nodes should I disrupt in order to partition the network?") or which node was the source of a given transaction (to remove pseudonymity). This PR means that instead of actively probing via the above method, you could also actively probe by sending "GETDATA" messages for txids you haven't yet received an "INV" for.

sr-gi commented:
> The easy way to probe for the presence of transaction X in the mempool is to propose a fake transaction Z that spends an unspent output of X, and a fake transaction Y, with fake signatures for both, and see how its orphan handling behaves. If you receive a request for X and Y, then the victim didn't have X in its mempool; if you receive a request for just Y, then it did have X in the mempool; if you don't receive a request for either, then X was not in the mempool and was in its reject filter.

This feels pretty similar to what [TxProbe](https://arxiv.org/abs/1812.00942) used to do to infer links between nodes, and ultimately relates to how orphans are treated differently from mempool transactions both when handed and requested. 

This is pretty tangential to this PR, but I'm wondering whether it may make sense to change that behavior to always request all the unannounced parents on a given link when an orphan is received through it. In the same way, orphans could be requested, even if known, when requested through a link they haven't been received from. Validation for this type of transaction may even be simplified, given they are already known, so as long as the same data is received nothing else needs to be checked. At a glance, this looks cheap while it will prevent leaking information related to orphans.

jonatack reviewed (commented):
Initial concept/approach ACK 500e31e87b33418a0747a0f3118db9ceb23896be

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -83,6 +84,7 @@ class CTxMemPoolEntry
     const int64_t sigOpCost;        //!< Total sigop cost
     CAmount m_modified_fee;         //!< Used for determining the priority of the transaction for mining in a block
     LockPoints lockPoints;          //!< Track the height and time at which tx was final
+    const SteadyClock::time_point m_entry_time; //!< High-precision time when entering the mempool
```
No longer relevant

ajtowns commented:
Rebased and switched to using the mempool sequence instead of `SteadyClock`

jonatack reviewed (commented):
Approach ACK with some questions and feedback.

The last push changed from storing the entry time to using the sequence number in `mempool#GetSequence` per https://github.com/bitcoin/bitcoin/pull/27675#discussion_r1230714110.

The `CTxMemPool::m_sequence_number` documentation could use updating in this case regarding its expanded role

```cpp
    // In-memory counter for external mempool tracking purposes.
    // This number is incremented once every time a transaction
    // is added or removed from the mempool for any reason.
    mutable uint64_t m_sequence_number GUARDED_BY(cs){1};
```

and this part of the PR description updated: `tracking the timestamp of the last INV message and comparing that against the mempool entry time`.

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -120,6 +123,13 @@ class CTxMemPoolEntry
+
+    // can't be moved
+    CTxMemPoolEntry(CTxMemPoolEntry&&) = default;
+    CTxMemPoolEntry& operator=(CTxMemPoolEntry&&) = delete;
```
Oh, yuck; that's a bad rebase. :(

ajtowns commented on src/net_processing.cpp:
```diff
@@ -2290,22 +2282,14 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
-        if ((mempool_req.count() && txinfo.m_time <= mempool_req) || txinfo.m_time <= now - UNCONDITIONAL_RELAY_DELAY) {
-            return std::move(txinfo.tx);
-        }
+        return std::move(txinfo.tx);
```
That's moving a member of a local struct/object, not a simple local, so it can't have been constructed in place and copy/move elision isn't possible, though?

ajtowns replied:
I think the test you want is `auto s = Wrapper(); return std::move(s.snitch);`

ajtowns commented on src/net_processing.cpp:
```diff
@@ -5781,30 +5739,35 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                         }
                         tx_relay->m_tx_inventory_known_filter.insert(hash);
-                        if (hash != txid) {
+                        if (peer->m_wtxid_relay && hash != txinfo.tx->GetHash()) {
```
For wtxid relay peers the prior `.insert(hash)` line adds the wtxid to the filter; for non-wtxid relay peers, it adds the txid to the filter. We only need to do a second insert to add the txid for wtxid relay peers, and then only when the wtxid is actually different to the txid. The extra check is just because testing a bool is easier than comparing two uint256s.

glozow replied:
Is adding the txid still necessary though?

IIUC, the purpose of adding this transaction's txid to `inventory_known_filter` is to save a step if this transaction has an unconfirmed child (if it already existed in this filter, then we wouldn't add it to the `recently_announced` filter). The PR is deleting those lines from `ProcessGetData`, and that's fine since an unconfirmed parent's sequence number should before that of its child.
But I'm not able to see another reason why we should add the txid to the `inventory_known_filter`, so why not delete this whole block?

If it's still needed, it probably makes sense to update the 2 comments since they mention recently_announced filter.

ajtowns replied:
Maybe we could recover that as an invariant by having txs added in `MaybeUpdateMempoolForReorg` set the sequence number as 0? The idea being that if they've made it into a block then we've probably announced the headers already so it's no secret that we knew about the tx anyway...

ajtowns replied:
> But I'm not able to see another reason why we should add the txid to the `inventory_known_filter`, so why not delete this whole block?

That seems correct to me.

ajtowns replied:
Updated to not add txids to the known filter for wtxid peers, and to set the entry sequence to 0 when adding txs that were in a block that was reorged out.

naumenkogs replied:
This is done in the following commit [c88deec](https://github.com/bitcoin/bitcoin/pull/27675/commits/c88deec99d53f55ff76fedaa30c318be6b72d085)

ajtowns commented on src/net_processing.cpp:
```diff
@@ -5781,30 +5739,35 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
+                    // in m_tx_inventory_to_send and INVed above with a mempool nTime after
+                    // current_time. As such, only get the current time after sending all the invs.
+                    LOCK(m_mempool.cs);
+                    tx_relay->m_last_inv_sequence = m_mempool.GetSequence();
```
Updated

naumenkogs replied:
34eab37b08bb9f23459aad179685a505a46759d7
Seems to be still talking about `current_time`?

ajtowns replied:
I think I messed this up in the [bad rebase](https://github.com/bitcoin/bitcoin/pull/27675#discussion_r1253703097) :( Should be fixed now.

ajtowns commented:
Updated for jonatack's review -- removes unneeded code changes, updates comments.

jonatack reviewed (commented):
ACK 172fa815e08b54c61b24d552f065df15046aef3c

With this simplification, test coverage for the new logic might be feasible?

glozow commented on src/net_processing.cpp:
```diff
@@ -147,8 +147,12 @@ static constexpr auto OUTBOUND_INVENTORY_BROADCAST_INTERVAL{2s};
 /** Maximum rate of inventory items to send per second.
  *  Limits the impact of low-fee transaction floods. */
 static constexpr unsigned int INVENTORY_BROADCAST_PER_SECOND = 7;
+/** Target number of inventory items to send per transmission. */
```
```suggestion
/** Target number of transaction inventory items to send per transmission. */
```

ajtowns committed:
net_processing: don't add txids to m_tx_inventory_known_filter

We no longer have m_recently_announced_invs, so there is no need to add
txids to m_tx_inventory_known_filter to dedupe that filter.

ajtowns committed:
validation: when adding txs due to a block reorg, allow immediate relay

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -130,6 +132,7 @@ class CTxMemPoolEntry
     int32_t GetTxWeight() const { return nTxWeight; }
     std::chrono::seconds GetTime() const { return std::chrono::seconds{nTime}; }
     unsigned int GetHeight() const { return entryHeight; }
+    uint64_t GetSequence() const { return entry_sequence; }
```
It's okay to discard it though -- you just might as well not have called the function in the first place if you don't care about the return value? (And given it's inline, presumably the compiler will treat it as if it wasn't called anyway).

ajtowns replied:
I can see the point of adding `nodiscard` in that case for confusing APIs (eg, where someone not familiar with the API might write `foo.empty()` when they meant `foo.clear()`) but for a `foo.GetBar()` function, it just seems like busy work, that I don't think makes sense to do.

naumenkogs commented:
ACK 0753120b6cb270914753d30a2626272f896a8545 modulo [this (probably lost?) comment](https://github.com/bitcoin/bitcoin/pull/27675#discussion_r1265162383)

My favorite achievement of this PR is dropping the complication of associated logic :)

ajtowns commented on src/validation.cpp:
```diff
@@ -834,7 +834,10 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, m_active_chainstate.m_chain.Height(), m_pool.GetSequence(),
+    // Set entry_sequence to 0 when bypass_limits is used; this allows txs from a block
+    // reorg to be marked earlier than any child txs that were already in the mempool.
+    const uint64_t entry_sequence = args.m_bypass_limits ? 0 : m_pool.GetSequence();
```
Dropped

naumenkogs commented:
ACK f5d17e1a3f2856483ad3a6abcf3ea415f2d05fde

ajtowns commented:
Updated to remove some unnecessary constructor changes and misleading comments associated with them.

naumenkogs commented:
ACK 7efb30b

glozow committed:
test: Check tx from disconnected block is immediately requestable

Check that peers can immediately request txs from blocks that have been
reorged out and are now in our mempool.

glozow reviewed (commented):
Approach ACK, mempool sequence seems suitable for this.

I think we should add a functional test, perhaps [something like this](https://github.com/glozow/bitcoin/commit/258b1c8ced55b528d1afc585c23a3d98fa30394b)

ajtowns commented on src/validation.cpp:
```diff
@@ -834,7 +834,10 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, m_active_chainstate.m_chain.Height(),
+    // Set entry_sequence to 0 when bypass_limits is used; this allows txs from a block
+    // reorg to be marked earlier than any child txs that were already in the mempool.
+    const uint64_t entry_sequence = bypass_limits ? 0 : m_pool.GetSequence();
```
> Doesn't that mean there are races? (There is no guarantee that `TransactionAddedToMempool()` will always run in the near future)?

The mempool value is incremented when TrAddedTMP is queued, not when it's eventually run -- the increment happens first, then the result of the increment is passed through and queued up. So I don't think there's any racy here.

> Perhaps it makes sense to use `m_pool.GetSequence() + 1` here?

I think it would make more sense to change `info_for_relay`?

```c++
TxMempoolInfo CTxMemPool::info_for_relay(const GenTxid& gtxid, uint64_t last_sequence) const
{
    LOCK(cs);
    indexed_transaction_set::const_iterator i = (gtxid.IsWtxid() ? get_iter_from_wtxid(gtxid.GetHash()) : mapTx.find(gtxid.GetHash()));
//  if (i == mapTx.end() || last_sequence < i->GetSequence()) {
    if (i == mapTx.end() || last_sequence <= i->GetSequence()) {
        return TxMempoolInfo();
    } else {
        return GetInfo(i);
    }
}
```

glozow replied:
> Doesn't that mean there are races? (There is no guarantee that TransactionAddedToMempool() will always run in the near future)?

I meant `TransactionAddedToMempool` event happening, not the validation interface clients' callbacks. IIUC the event firing on main signals (which calls `GetAndIncrementSequence()`) happens in validation before `ProcessTransaction` returns:
https://github.com/bitcoin/bitcoin/blob/e5a9f2fb62dc4db6cad21b2997d96a881ea64125/src/validation.cpp#L1234

The queued callbacks are scheduled and may not have necessarily happened yet, but PeerManager doesn't implement `TransactionAddedToMempool`. It is only interested in the current `GetSequence()` and what is recorded in `mempool_sequence`.

> Basically, I am wondering what happens if you reorder t=2 and t=3 in your example.

My understanding is that `GetSequence()` returns 2 when `ProcessTransaction` is done, even if t=2 and t=3 are swapped.

glozow replied:
Ah I didn't see the last message.

> I think it would make more sense to change info_for_relay?

sounds good to me

ajtowns replied:
Updated. This also changes the initial value for `m_last_inv_sequence` to `1` instead of `0`. (Note that I reversed the order of the `if .. else` in `info_for_relay` so `GetSeq() < last` becomes `! (GetSeq() <= last)` which is `last < GetSeq()`)

ajtowns commented:
Update for glozow's review; fixes off-by-one error and adds some glozow's test coverage of reorg behaviour

naumenkogs commented on src/kernel/mempool_entry.h:
```diff
@@ -79,6 +79,7 @@ class CTxMemPoolEntry
     const size_t nUsageSize;        //!< ... and total memory usage
     const int64_t nTime;            //!< Local time when entering the mempool
     const unsigned int entryHeight; //!< Chain height when entering the mempool
+    const uint64_t entry_sequence;  //!< Sequence number used to determine w hether this transaction is too recent for relay
```
1e9684f39fba909b3501e9402d5b61f4bf744ff2
"w hether"

naumenkogs reviewed (approved):
ACK 1e9684f39fba909b3501e9402d5b61f4bf744ff2

glozow reviewed (commented):
ACK 1a118062fbc4ec8f645f4ec4298d869a869c3344

mzumsande reviewed (commented):
My understanding is that there is a bit of a trade-off with respect to memory (correct me if the byte counting is off!):
While it saves 33kB per peer (so ~4MB with 125 peers; ~250kB for non-listening nodes with 8 peers), it also adds at least 8 bytes per mempool entry for `uint64_t entry_sequence`.
(for me, `sizeof(CTxMemPoolEntry)` even grows from 264 to 280 bytes - packing effects?).
This seems a bit unsatisfactory because the `entry_sequence` data won't be needed after a few seconds (i.e. after we've sent an inv to each peer).

Assuming a full mempool with `~150k` mempool entries, that is between 1.2 and 2.4MB.
Because this is included in the maxmempool limit, it's effectively a decrease in mempool capacity instead of additional memory when the mempool is full.

I wonder if there would be a more memory-efficient way to achieve the same behavior - did you consider something like a global list of all recent transactions with timestamp, from which entries could get removed regularly (i.e., after we have sent an inv to each tx-relay peer).

ajtowns committed:
mempool_entry: improve struct packing

ajtowns commented:
> (for me, `sizeof(CTxMemPoolEntry)` even grows from 264 to 280 bytes - packing effects?)

Added a commit to fix this.

> My understanding is that there is a bit of a trade-off with respect to memory (correct me if the byte counting is off!): While it saves 33kB per peer (so ~4MB with 125 peers; ~250kB for non-listening nodes with 8 peers), it also adds at least 8 bytes per mempool entry for `uint64_t entry_sequence`. [..] This seems a bit unsatisfactory because the `entry_sequence` data won't be needed after a few seconds (i.e. after we've sent an inv to each peer).
>
> Assuming a full mempool with `~150k` mempool entries, that is between 1.2 and 2.4MB. Because this is included in the maxmempool limit, it's effectively a decrease in mempool capacity instead of additional memory when the mempool is full.

Effectively, that's a 0.4% increase in memory usage per tx (2000B per tx for 150k txs in 300MB becomes 2008B per tx for 149.4k txs in 300MB).

The main point here though is that with this change is for the case where we need to increase the number of (tx relay) peers or the tx acceptance rate -- the former meaning we'd need to allocate more bloom filters, and the latter meaning we'd need each bloom filter to be significantly bigger (cf #27630).

> I wonder if there would be a more memory-efficient way to achieve the same behavior - did you consider something like a global list of all recent transactions with timestamp, from which entries could get removed regularly (i.e., after we have sent an inv to each tx-relay peer).

I did think about that, but it seemed significantly more complicated -- you'd need to either have a map of txid-to-sequence, or a map of txiter-to-sequence; the former would be wasting 24B per entry duplicating the txid, but the latter would need a lot of care to make sure there aren't dangling references to txiters that have been removed from the mempool (and may be reallocated). You'd also need to regularly remove things from this map based on something like `min(peer.last_inv)` over each peer, which seems slightly awkward -- the information in the mempool starts depending on p2p state, rather than just the txs that are in the mempool. 

I do think those approaches would both result in measurable memory savings (1.2MB down to perhaps 100kB at 22tx/sec over 60 seconds?), but to me it just doesn't seem worth the complexity overall.

amitiuttarwar reviewed (commented):
review ACK fb02ba3c5f5

it's been a while since I've hung out with mempool transaction code, so in addition to reading through the code I spent a good chunk of energy refreshing contextual behaviors. some context behind the ack: 
- thinking through if there is a meaningful shift in privacy model: I find these changes acceptable
- understanding the memory tradeoff: this approach makes sense to me, esp with #27630 in the pipeline
- re-familiarizing myself with how transactions enter the mempool
- understanding how the mempool sequence number is handled 
- thinking through edge cases & understanding the ones flagged in prior review of this PR

tangential to this PR, but sharing since I noticed- `m_sequence_number` could be private and non `mutable` by removing the `const` from `GetAndIncrementSequence` (which doesn't conceptually make sense anyway)

ajtowns commented on src/test/util/txmempool.h:
```diff
@@ -30,6 +31,7 @@ struct TestMemPoolEntryHelper {
     TestMemPoolEntryHelper& Fee(CAmount _fee) { nFee = _fee; return *this; }
     TestMemPoolEntryHelper& Time(NodeSeconds tp) { time = tp; return *this; }
     TestMemPoolEntryHelper& Height(unsigned int _height) { nHeight = _height; return *this; }
+    TestMemPoolEntryHelper& Sequence(uint64_t _seq) { m_sequence = _seq; return *this; }
```
No, it's not

naumenkogs commented:
ACK fb02ba3c5f5bcd96b5e3622ef001b8e57ce63fc0

glozow reviewed (commented):
reACK fb02ba3c5f5bcd96b5e3622ef001b8e57ce63fc0

maflcko reviewed (commented):
lgtm, one doc nit.


Approach ACK fb02ba3c5f5bcd96b5e3622ef001b8e57ce63fc0, though I have not closely reviewed e4ffabbffacc4b890d393aafcc8286916ef887d8 🏁

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: Approach ACK fb02ba3c5f5bcd96b5e3622ef001b8e57ce63fc0, though I have not closely reviewed e4ffabbffacc4b890d393aafcc8286916ef887d8 🏁
wZjhZp82DT5gvd1V2EbD+Lsejn1GZvm9++6dmmXGfe1v+61gbP3gMtecA9Vz94YNZ3KT8EyzjbMr83Tiw/rEBg==
```

</details>

ajtowns commented on src/validation.cpp:
```diff
@@ -834,7 +834,10 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, m_active_chainstate.m_chain.Height(), m_pool.GetSequence(),
+    // Set entry_sequence to 0 when bypass_limits is used; this allows txs from a block
+    // reorg to be marked earlier than any child txs that were already in the mempool.
+    const uint64_t entry_sequence = bypass_limits ? 0 : m_pool.GetSequence();
```
Unique for any tx and strictly monotonically increasing for child txs would also be violated for accepting packages to the mempool, I think.

glozow replied:
Every tx submitted together in `SubmitPackage` has the same sequence number yes. But it should hold that if x is spent by y then x.sequence <= y.sequence.
